### PR TITLE
Add PINO layers and layer tests

### PR DESCRIPTION
**Pull request by DARHWOLF**
<h2>Description</h2>
<p> Referenced from PR #4415 </p>
<p>
  This PR introduces new <strong>Physics-Informed Neural Operator (PINO)</strong> layers to DeepChem, designed to enhance the solution of partial differential equations (PDEs) by leveraging physics-informed loss functions and boundary condition enforcement. These layers are based on the approach outlined in 
  <a href="https://arxiv.org/abs/2111.03794" target="_blank">
    Physics-Informed Neural Operator for Learning Partial Differential Equations
  </a>.
</p>

<p>
  The added layers include:
  <ul>
    <li><strong>ActNorm</strong>: A learnable affine normalization layer that applies scaling and bias.</li>
    <li><strong>FourierBlock1D</strong>: A 1D Fourier spectral convolution block that retains a specified number of Fourier modes.</li>
  </ul>
  These layers will serve as building blocks for PINO-based models in DeepChem, with a focus on solving PDEs like the 1D Burgers' equation. The Fourier block is designed to efficiently capture function representations in the frequency domain, while the ActNorm layer ensures proper normalization during training.
</p>

<p>
  Comprehensive unit tests have been added to ensure:
</p>

<ul>
  <li>Correct initialization and functionality of the PINO layers</li>
  <li>Proper gradient flow through the FourierBlock1D layer</li>
  <li>Accurate application of mode truncation in the FourierBlock1D layer</li>
  <li>Ensuring zero weights result in zero output, verifying the identity behavior of the layers</li>
  <li>Consistency with NumPy FFT results for numerical accuracy</li>
</ul>

<p> 

- [X] The code has been thoroughly tested in a local Python 3.11 environment. All unit tests were executed using <code>pytest</code>, and the **test suite completed successfully** without errors. 

</p>

<h2>Type of change</h2>

<ul>
  <li>[ ] Bug fix (non-breaking change which fixes an issue)</li>
  <li>[x] New feature (non-breaking change which adds functionality)</li>
  <li>[ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)</li>
  <li>[X] Documentation (modification for documents)</li>
</ul>

<h2>Checklist</h2>

<ul>
  <li>[x] My code follows <a href="https://deepchem.readthedocs.io/en/latest/development_guide/coding.html">the style guidelines of this project</a></li>
  <li>[x] Run <code>yapf -i deepchem/models/torch_models/layers.py</code> and <code>yapf -i deepchem/models/torch_models/tests/test_pino_layers.py</code> (version 0.32.0)</li>
  <li>[x] Run <code>mypy -p deepchem</code> and check no errors (ignoring third-party stub issues)</li>
  <li>[x] Run <code>flake8 deepchem/models/torch_models/layers.py</code> and <code>flake8 deepchem/models/torch_models/tests/test_pino_layers.py --count</code></li>
  <li>[x] Run <code>python -m doctest deepchem/models/torch_models/layers.py</code> and check no errors</li>
  <li>[x] I have performed a self-review of my own code</li>
  <li>[x] I have commented my code, particularly in hard-to-understand areas</li>
  <li>[x] I have made corresponding changes to the documentation</li>
  <li>[x] I have added tests that prove my feature works</li>
  <li>[x] New unit tests pass locally with my changes</li>
  <li>[x] I have checked my code and corrected any misspellings</li>
</ul>

### CODE REVIEW

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -432,6 +432,107 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
 
+class ActNorm(nn.Module):
+    """
+    PINO Activation normalization layer that applies learnable scaling and bias.
+
+    This layer is a per-channel affine normalization used by the
+    Physics-Informed Neural Operator (PINO) to stabilize spectral
+    block outputs. It applies the learnable transformation:
+
+        y = scale * x + bias
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:443]: Does this render correct? You may want to try rendering the docs. I am not sure if this equation displays

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -432,6 +432,107 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+    ----------
+    num_features : int
+        Number of features/channels in the input.
+
+    Examples
+    --------
+    >>> norm = ActNorm(dim=64)
+    >>> x = torch.randn(10, 1, 64)
+    >>> y = norm(x)
+    >>> print(y.shape)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:455]: Can you cite a source?

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -432,6 +432,107 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        super().__init__()
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.modes = modes
+        self.scale = 1 / (in_channels * out_channels)**0.5
+        self.weights = nn.Parameter(
+            self.scale * torch.rand(in_channels, out_channels, modes, 2))
+        self.norm = ActNorm(out_channels)
+
+    def compl_mul1d(self, a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:508]: Add a docstring here please for this method

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -432,6 +432,107 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        real = torch.einsum("bci,oci->boi",
+                            a[..., 0], b[..., 0]) - torch.einsum(
+                                "bci,oci->boi", a[..., 1], b[..., 1])
+        imag = torch.einsum("bci,oci->boi",
+                            a[..., 0], b[..., 1]) + torch.einsum(
+                                "bci,oci->boi", a[..., 1], b[..., 0])
+        result = torch.stack([real, imag], dim=-1)
+        return result
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:518]: Add a docstring

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -432,6 +432,107 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        y = scale * x + bias
+
+    Parameters
+    ----------
+    num_features : int
+        Number of features/channels in the input.
+
+    Examples
+    --------
+    >>> norm = ActNorm(dim=64)
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/layers.py:452]: Please add an import line: "from deepchem.models.torch_models.layers import ActNorm"

File: deepchem/models/torch_models/layers.py [hunk 6]
```diff
@@ -432,6 +432,107 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+    >>> print(y.shape)
+    torch.Size([10, 1, 64])
+    """
+
+    def __init__(self, dim: int):
+        super().__init__()
+        self.scale = nn.Parameter(torch.ones(1, 1, dim))
+        self.bias = nn.Parameter(torch.zeros(1, 1, dim))
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/layers.py:464]: Can you please add brief docstrings explaining the inputs, outputs and shapes?

File: deepchem/models/torch_models/layers.py [hunk 7]
```diff
@@ -432,6 +432,107 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+    in_channels : int
+        Number of input channels.
+    out_channels : int
+        Number of output channels.
+    modes : int
+        Number of Fourier modes to retain in the frequency domain.
+
+    Examples
+    --------
+    >>> block = FourierBlock1D(in_channels=64, out_channels=64, modes=16)
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/layers.py:491]: Please add an import of FourierBlock1D

File: deepchem/models/torch_models/layers.py (hunk 5 above)
Comment (DARHWOLF) [reply deepchem/models/torch_models/layers.py:491]: Added the import line: "from deepchem.models.torch_models.layers import ActNorm".

File: deepchem/models/torch_models/layers.py (hunk 6 above)
Comment (DARHWOLF) [reply deepchem/models/torch_models/layers.py:505]: Added docstrings explaining the inputs, outputs and shapes.

File: deepchem/models/torch_models/layers.py (hunk 7 above)
Comment (DARHWOLF) [reply deepchem/models/torch_models/layers.py:575]: Added import of FourierBlock1D

File: deepchem/models/torch_models/layers.py (hunk 2 above)
Comment (DARHWOLF) [reply deepchem/models/torch_models/layers.py:496]: Added

File: deepchem/models/torch_models/layers.py (hunk 3 above)
Comment (DARHWOLF) [reply deepchem/models/torch_models/layers.py:592]: Added docstrings for this method

File: deepchem/models/torch_models/layers.py (hunk 4 above)
Comment (DARHWOLF) [reply deepchem/models/torch_models/layers.py:624]: Done
